File: 990-satisfiability-of-equality-equations/990-satisfiability-of-equality-equations.js
/**
 * @param {string[]} equations
 * @return {boolean}
 */
var equationsPossible = function(equations) {
  const parentOf = {};
  for (const char of 'abcdefghijklmnopqrstuvwxyz') {
    parentOf[char] = char;
  }
  
  const find = function (char) {
    return parentOf[char] === char ? char : find(parentOf[char]);
  }
  
  // Iterate over equations and form unions
  for (const eq of equations) {
    if (eq[1] === '=') {
      const x = eq[0];
      const y = eq[3];
      parentOf[find(x)] = find(y);
    }
  }
  
  // iterate over equations and check for conflicts
  for (const eq of equations) {
    if (eq[1] === '!') {
      const x = eq[0];
      const y = eq[3];
      if (parentOf[find(x)] === find(y)) {
        return false;
      }
    }
  }
  
  return true
};

File: 990-satisfiability-of-equality-equations/990-satisfiability-of-equality-equations.py
class Solution:
    def equationsPossible(self, equations: List[str]) -> bool:
        parent_of = { c: c for c in ascii_lowercase }
        
        def find(char: str) -> str:
            """Search the unions for the top-level element.
            The commented line runs a union find without path compression.
            """
            # return char if parent_of[char] == char else find(parent_of[char])
            if parent_of[char] == char:
                return char
            parent_of[char] = find(parent_of[char])
            return parent_of[char]
        
        # Iterate over equations and form unions
        for x, eq, _, y in equations:
            if eq == "=":
                parent_of[find(x)] = find(y)
        
        # Iterate over equations and check for conflicts
        for x, eq, _, y in equations:
            if eq == "!" and parent_of[find(x)] == find(y):
                return False
        
        return True